replica/table.cc

```diff
@@ -814,6 +797,43 @@ future<> table::parallel_foreach_compaction_group(std::function<future<>(compact
     });
 }
 
+static std::pair<dht::token, dht::token> get_token_range(const dht::partition_range& pr) {
+    assert(!pr.is_singular());
+    auto start_token = pr.start() ? pr.start()->value().token() : dht::first_token();
+    auto end_token = pr.end() ? pr.end()->value().token() : dht::maximum_token();
+    return std::make_pair(start_token, end_token);
+}
+
+stop_iteration table::foreach_compaction_group_until(const dht::partition_range& pr, std::function<stop_iteration(compaction_group&)> action) {
+    if (pr.is_singular()) {
+        return action(compaction_group_for_token(pr.start()->value().token()));
+    }
+    auto [start_token, end_token] = get_token_range(pr);
+    for (auto token = start_token; token < end_token; ) {
+        auto& cg = compaction_group_for_token(token);
+        token = cg.next_compaction_group_token();
+        if (action(cg) == stop_iteration::yes) {
+            return stop_iteration::yes;
+        }
+    }
+    return stop_iteration::no;
+}
+
+future<stop_iteration> table::foreach_compaction_group_gently_until(const dht::partition_range& pr, std::function<future<stop_iteration>(compaction_group&)> action) {
+    if (pr.is_singular()) {
+        co_return co_await action(compaction_group_for_token(pr.start()->value().token()));
+    }
+    auto [start_token, end_token] = get_token_range(pr);
+    for (auto token = start_token; token < end_token; ) {
+        auto& cg = compaction_group_for_token(token);
+        token = cg.next_compaction_group_token();
+        if (co_await action(cg) == stop_iteration::yes) {
+            co_return stop_iteration::yes;
+        }
+    }
+    co_return stop_iteration::no;
```

Comment:
What protects against concurrent changes?

Reply:
Currently, nothing. it's up to the caller to get what it needs from the compaction_group synchronously.
When we inplement split end to end we'll need something like a rwlock to protect storage groups so they can be updated atomically.